### PR TITLE
Re enable HttpMetricsHandlerTests in netty5 branch

### PR DESCRIPTION
**Pull request by pderop**
Related to #1873

This PR is an attempt to bring back the HttpMetricsHandlerTests into netty5 branch.
mainly, here what has been done:

- in AbstractHttpClientMetricsHandler.java: it seems that now the processed data must also be extracted from the HttpContent parameter that can be passed as argument. I think that the ByteBufHolder is not used anymore, but I did not remote the test on it.
- Same thing done for AbstractHttpServerMetricsHandler.java: the HttpContent must be checked from the extractProcessedDataFromBuffer method.
- in HttpTrafficHandler.java, it seems that the handler can't fire an event after having removed itself from the pipeline, so the removal from the pipeline is now done after the event is fired. [See netty DefaultChannelHandlerContext.java](https://github.com/netty/netty/blob/main/transport/src/main/java/io/netty5/channel/DefaultChannelHandlerContext.java#L347-L355)
- In AbstractHttpServerMetricsHandler.write method, sometimes the ChannelOperations seems to not be available anymore during listener invocation. This is probably caused by this [netty issue](https://github.com/netty/netty/pull/9489), where listeners are now invoked asynchronously. So the ChannelOperations is retrieved and cached before the listener is added.
- HttpMetricsHandlerTests.java: the test has been refactored in order to always expect to observe 4 disconnects when H1 is used, and three disconnects when H2 or H2C is used.

### CODE REVIEW

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpTrafficHandler.java [hunk 1]
```diff
@@ -92,7 +93,7 @@ else if (evt == UPGRADE_SUCCESSFUL) {
 				log.debug(format(channel, "The upgrade to H2C protocol was successful."));
 			}
 			sendNewState(Connection.from(channel), HttpClientState.UPGRADE_SUCCESSFUL);
-			ctx.pipeline().remove(this);
+			removeThisHandler = true; // we have to remove ourself from the pipleline after having fired the event below.
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/HttpTrafficHandler.java:96]: nit:

```suggestion
			removeThisHandler = true; // we have to remove ourselves from the pipeline after having fired the event below.
```

**Comment by pderop** [comment]
please do not review for the moment, some tests are failing.

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 2]
```diff
@@ -171,8 +169,9 @@ void tearDown() {
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testExistingEndpoint(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
-		CountDownLatch latch1 = new CountDownLatch(4); // expect to observe 2 server disconnect + 2 client disconnect events
+			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negociatedProtocol) throws Exception {
+		int expectedCloses = getExpectedCloses(negociatedProtocol);
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:173]: nit:

```suggestion
			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negotiatedProtocol) throws Exception {
		int expectedCloses = getExpectedCloses(negotiatedProtocol);
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 3]
```diff
@@ -241,7 +240,8 @@ else if (clientProtocols.length == 2 &&
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testRecordingFailsServerSide(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) {
+	                                  @Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx,
+	                                  @SuppressWarnings("unused") HttpProtocol negociatedProtocol) {
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:244]: nit:

```suggestion
	                                  @SuppressWarnings("unused") HttpProtocol negotiatedProtocol) {
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 4]
```diff
@@ -265,7 +265,8 @@ void testRecordingFailsServerSide(HttpProtocol[] serverProtocols, HttpProtocol[]
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testRecordingFailsClientSide(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) {
+	                                  @Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx,
+	                                  @SuppressWarnings("unused") HttpProtocol negociatedProtocol) {
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:269]: nit:

```suggestion
	                                  @SuppressWarnings("unused") HttpProtocol negotiatedProtocol) {
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 5]
```diff
@@ -287,11 +288,10 @@ void testRecordingFailsClientSide(HttpProtocol[] serverProtocols, HttpProtocol[]
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testNonExistingEndpoint(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
+			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negociatedProtocol) throws Exception {
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:291]: nit:

```suggestion
			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negotiatedProtocol) throws Exception {
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 6]
```diff
@@ -287,11 +288,10 @@ void testRecordingFailsClientSide(HttpProtocol[] serverProtocols, HttpProtocol[]
...
 		// For HTTP11, we expect to observe 2 DISCONNECTS for client, and 2 DISCONNECT for server.
 		// Else, we expect to observe 2 DISCONNECTS for client, and 1 DISCONNECT for server.
-		boolean isHTTP11 = clientProtocols.length == 1 && clientProtocols[0] == HttpProtocol.HTTP11;
-		int expectedDisconnects = isHTTP11 ? 4 : 3;
+		int expectedDisconnects = getExpectedCloses(negociatedProtocol);
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:294]: nit:

```suggestion
		int expectedDisconnects = getExpectedCloses(negotiatedProtocol);
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 7]
```diff
@@ -369,8 +369,9 @@ else if (protocols.contains(HttpProtocol.H2) || protocols.contains(HttpProtocol.
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testUriTagValueFunction(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
-		CountDownLatch latch1 = new CountDownLatch(4); // expect to observe 2 server disconnect + 2 client disconnect events
+			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negociatedProtocol) throws Exception {
+		int expectedCloses = getExpectedCloses(negociatedProtocol);
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:373]: nit:

```suggestion
			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negotiatedProtocol) throws Exception {
		int expectedCloses = getExpectedCloses(negotiatedProtocol);
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 8]
```diff
@@ -420,8 +421,9 @@ else if (clientProtocols.length == 2 &&
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testUriTagValueFunctionNotSharedForClient(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
-		CountDownLatch latch1 = new CountDownLatch(4); // expect to observe 2 server disconnect + 2 client disconnect events
+			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negociatedProtocol) throws Exception {
+		int expectedCloses = getExpectedCloses(negociatedProtocol);
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:425]: nit:

```suggestion
			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negotiatedProtocol) throws Exception {
		int expectedCloses = getExpectedCloses(negotiatedProtocol);
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 9]
```diff
@@ -499,7 +501,8 @@ else if (clientProtocols.length == 2 &&
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testContextAwareRecorderOnClient(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
+	                                      @Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx,
+	                                      @SuppressWarnings("unused") HttpProtocol negociatedProtocol) throws Exception {
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:505]: nit:

```suggestion
	                                      @SuppressWarnings("unused") HttpProtocol negotiatedProtocol) throws Exception {
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 10]
```diff
@@ -530,7 +533,8 @@ void testContextAwareRecorderOnClient(HttpProtocol[] serverProtocols, HttpProtoc
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testContextAwareRecorderOnServer(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
+	                                      @Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx,
+	                                      @SuppressWarnings("unused") HttpProtocol negociatedProtocol) throws Exception {
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:537]: nit:

```suggestion
	                                      @SuppressWarnings("unused") HttpProtocol negotiatedProtocol) throws Exception {
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 11]
```diff
@@ -562,8 +566,9 @@ void testContextAwareRecorderOnServer(HttpProtocol[] serverProtocols, HttpProtoc
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testServerConnectionsMicrometer(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
-		CountDownLatch latch = new CountDownLatch(4); // expect to observe 2 server disconnect + 2 client disconnect events
+			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negociatedProtocol) throws Exception {
+		int expectedCloses = getExpectedCloses(negociatedProtocol);
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:570]: nit:

```suggestion
			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx, HttpProtocol negotiatedProtocol) throws Exception {
		int expectedCloses = getExpectedCloses(negotiatedProtocol);
```

File: reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java [hunk 12]
```diff
@@ -618,7 +623,8 @@ void testServerConnectionsMicrometer(HttpProtocol[] serverProtocols, HttpProtoco
 	@ParameterizedTest
 	@MethodSource("httpCompatibleProtocols")
 	void testServerConnectionsRecorder(HttpProtocol[] serverProtocols, HttpProtocol[] clientProtocols,
-			@Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx) throws Exception {
+	                                   @Nullable ProtocolSslContextSpec serverCtx, @Nullable ProtocolSslContextSpec clientCtx,
+	                                   @SuppressWarnings("unused") HttpProtocol negociatedProtocol) throws Exception {
```
Comment (violetagg) [inline reactor-netty-http/src/test/java/reactor/netty/http/HttpMetricsHandlerTests.java:627]: nit:

```suggestion
	                                   @SuppressWarnings("unused") HttpProtocol negotiatedProtocol) throws Exception {
```

**Review by violetagg — CHANGES_REQUESTED** [review]

**Comment by pderop** [comment]
I have temporarily disabled the testRecordingFailsClientSide and testRecordingFailsClientSide tests, and now, we have successful tests for both ubuntu and macos. So, I tend to think that some more works need to be done for the other https://github.com/reactor/reactor-netty/pull/2335 PR, which is most likely to be again related (because if I disable the testRecordingFailsClientSide and the testRecordingFailsClientSide, then tests are passing OK).

Now, the remaining failing test is the same one for windows from the https://github.com/reactor/reactor-netty/pull/2338.

let's not merge this PR for the moment, I need to get back to the other https://github.com/reactor/reactor-netty/pull/2335.

File: reactor-netty-http/src/main/java/reactor/netty/http/client/AbstractHttpClientMetricsHandler.java [hunk 13]
```diff
@@ -165,6 +166,9 @@ private long extractProcessedDataFromBuffer(Object msg) {
 		else if (msg instanceof Buffer buffer) {
 			return buffer.readableBytes();
 		}
+		else if (msg instanceof HttpContent httpContent) {
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/AbstractHttpClientMetricsHandler.java:169]:
```
/.../reactor-netty/reactor-netty-http/src/main/java/reactor/netty/http/client/AbstractHttpClientMetricsHandler.java:169: warning: [rawtypes] found raw type: HttpContent
                else if (msg instanceof HttpContent httpContent) {
                                        ^
  missing type arguments for generic class HttpContent<R>
  where R is a type-variable:
    R extends HttpContent<R> declared in interface HttpContent
```

```suggestion
		else if (msg instanceof HttpContent<?> httpContent) {
```

File: reactor-netty-http/src/main/java/reactor/netty/http/server/AbstractHttpServerMetricsHandler.java [hunk 14]
```diff
@@ -208,6 +212,9 @@ private long extractProcessedDataFromBuffer(Object msg) {
 		else if (msg instanceof Buffer buffer) {
 			return buffer.readableBytes();
 		}
+		else if (msg instanceof HttpContent httpContent) {
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/server/AbstractHttpServerMetricsHandler.java:215]:
```
/.../reactor-netty/reactor-netty-http/src/main/java/reactor/netty/http/server/AbstractHttpServerMetricsHandler.java:215: warning: [rawtypes] found raw type: HttpContent
                else if (msg instanceof HttpContent httpContent) {
                                        ^
  missing type arguments for generic class HttpContent<R>
  where R is a type-variable:
    R extends HttpContent<R> declared in interface HttpContent
```

```suggestion
		else if (msg instanceof HttpContent<?> httpContent) {
```

**Review by violetagg — CHANGES_REQUESTED** [review]

File: reactor-netty-http/src/main/java/reactor/netty/http/server/AbstractHttpServerMetricsHandler.java (hunk 14 above)
Comment (pderop) [reply reactor-netty-http/src/main/java/reactor/netty/http/server/AbstractHttpServerMetricsHandler.java]: applied feedback, thanks !
I'm now investigating the unstable tests.

**Review by violetagg — APPROVED** [review]

**Comment by pderop** [comment]
Related to #1873

This PR is an attempt to bring back the HttpMetricsHandlerTests into netty5 branch.
What has been done:

- in AbstractHttpClientMetricsHandler.java: it seems that now the processed data must also be extracted from the HttpContent parameter that can be passed as argument.
- Same thing done for AbstractHttpServerMetricsHandler.java: the HttpContent must be checked from the extractProcessedDataFromBuffer method.
- in HttpTrafficHandler.java, it seems that the handler can't fire an event after having removed itself from the pipeline, so the removal from the pipeline is now done after the event is fired.
- In AbstractHttpServerMetricsHandler.write method, sometimes the ChannelOperations seems to not be available anymore during listener invocation. This is probably caused by this https://github.com/netty/netty/pull/9489, where listeners are now invoked asynchronously. So the ChannelOperations is retrieved and cached before the listener is added.
HttpMetricsHandlerTests.java: the test has been refactored in order to always expect to observe 4 disconnects when H1 is used, and three disconnects when H2 or H2C is used.
- The test is now rescheduling tasks in the event loops in order to ensure that meters are fully updated (remember that the meters are updated from listeners which are unfortunately rescheduled in event loops queue).

**Pull request merged**